pageserver/src/page_service.rs

```diff
@@ -752,16 +752,24 @@ impl PageServerHandler {
         }
 
         if request_lsn < **latest_gc_cutoff_lsn {
-            // Check explicitly for INVALID just to get a less scary error message if the
-            // request is obviously bogus
-            return Err(if request_lsn == Lsn::INVALID {
-                PageStreamError::BadRequest("invalid LSN(0) in request".into())
-            } else {
-                PageStreamError::BadRequest(format!(
+            let gc_info = &timeline.gc_info.read().unwrap();
+            if !gc_info.leases.contains_key(&request_lsn) {
+                // Check explicitly for INVALID just to get a less scary error message if the
+                // request is obviously bogus
+                return Err(if request_lsn == Lsn::INVALID {
+                    PageStreamError::BadRequest("invalid LSN(0) in request".into())
+                } else {
+                    PageStreamError::BadRequest(format!(
                         "tried to request a page version that was garbage collected. requested at {} gc cutoff {}",
                         request_lsn, **latest_gc_cutoff_lsn
                     ).into())
-            });
+                });
+            }
+            tracing::info!(
+                "requesting a leased lsn {} below gc cutoff {}",
```

Comment:
so every read request will print a line of info?

Reply:
...and having a read lock here seems costly, not sure if it is a good idea to change `latest_gc_cutoff_lsn` Rcu into `latest_gc_lease_cutoff_lsn`?

Reply:
> ...and having a read lock here seems costly, not sure if it is a good idea to change latest_gc_cutoff_lsn Rcu into latest_gc_lease_cutoff_lsn?

Are you talking about the `gc_info` read lock? We could do `latest_gc_lease_cutoff_lsn` but don't know what value it should be. It cannot be the lowest leased LSN because we don't know if everything between latest_gc_cutoff_lsn and latest_gc_lease_cutoff_lsn are valid.

Reply:
info log removal: 3c988d6666381f9aec6e565148b67a28399b29d1

Reply:
Like with the logging, we will only inspect the read lock when we are actually below the gc cutoff which is already rare, so I don't think the read lock taking will be significant.